File: packages/core/src/project.ts
/**
 * The projection: a derived, values-materialized, style-stripped view of a sheet, for full-text
 * search / RAG / an agent that only needs to *read* the data. It is never canonical, always
 * regenerated, and one-way — never parsed back as truth.
 */

import type { ComputedGrid } from './compute.js'
import { formatValue } from './format.js'
import { type Model, cloneModel, getCell } from './model.js'
import { serialize } from './serialize.js'
import { resolveStyles } from './styling.js'
import { type Locale, parseLiteral } from './values.js'

export interface ProjectOptions {
  computed?: ComputedGrid
  locale?: Locale
}

/** A model with formulas replaced by their computed display values and all styling removed. */
export function projectValuesModel(model: Model, opts: ProjectOptions = {}): Model {
  const next = cloneModel(model)
  for (const sheet of next.sheets) {
    const styles = resolveStyles(sheet)
    for (let r = 0; r < sheet.grid.length; r++) {
      for (let c = 0; c < sheet.width; c++) {
        const raw = sheet.grid[r]![c]!
        const attrs = styles.attrs(c, r + 1)
        if (raw.trim().startsWith('=')) {
          const v = opts.computed ? opts.computed.get(sheet.name, c, r + 1) : null
          sheet.grid[r]![c] = opts.computed
            ? formatValue(v, { format: attrs.format, locale: opts.locale })
            : raw
        } else if (attrs.format) {
          const v = parseLiteral(raw, opts.locale)
          if (typeof v === 'number') {
            sheet.grid[r]![c] = formatValue(v, { format: attrs.format, locale: opts.locale })
          }
        }
      }
    }
    sheet.styles = []
  }
  return next
}

/** Values-materialized markdown — the same clean table, but showing numbers instead of formulas. */
export function projectText(model: Model, opts: ProjectOptions = {}): string {
  return serialize(projectValuesModel(model, opts))
}

/**
 * A flat, prose-like projection: one line per data row, `header: value` pairs. Ideal as RAG chunk
 * units because each row is self-describing without the surrounding table structure.
 */
export function projectProse(model: Model, opts: ProjectOptions = {}): string {
  const lines: string[] = []
  for (const sheet of model.sheets) {
    const headers = sheet.grid[0] ?? []
    for (let r = 1; r < sheet.grid.length; r++) {
      const parts: string[] = []
      for (let c = 0; c < sheet.width; c++) {
        const header = headers[c]?.trim() || columnName(c)
        const raw = getCell(sheet, c, r + 1)
        const styles = resolveStyles(sheet)
        const attrs = styles.attrs(c, r + 1)
        let value: string
        if (raw.trim().startsWith('=')) {
          const v = opts.computed ? opts.computed.get(sheet.name, c, r + 1) : null
          value = opts.computed ? formatValue(v, { format: attrs.format, locale: opts.locale }) : raw
        } else {
          value = raw
        }
        if (value.trim() !== '') parts.push(`${header}: ${value}`)
      }
      if (parts.length) lines.push(`${sheet.name} — ${parts.join(', ')}`)
    }
  }
  return lines.join('\n')
}

function columnName(col: number): string {
  return `Col${col + 1}`
}

File: go/project.go
package defterparse

import (
	"strconv"
	"strings"
)

// The projection: a derived, style-stripped view of a model, for full-text
// search / RAG / an agent that only needs to *read* the data. It is never
// canonical, always regenerated, and one-way.
//
// TIER-1 BOUNDARY (Go pure-text layer): computed values are NOT materialized
// here. The reference TypeScript projectText, given a ComputedGrid, replaces
// formula cells with their evaluated display values and reformats literal
// numbers per each cell's `format` attribute. This Go layer deliberately has no
// formula engine and no number-format engine, so:
//   - a formula cell projects as its **source text** (leading `=` kept), and
//   - literal cells are emitted **verbatim** (no number-format application).
// Only the static style *rules* are stripped. Chart / conditional / validation /
// named-range declarations are retained (matching the reference, which clears
// only `sheet.styles`). Downstream, tela materializes computed values via a
// separate Node CLI.

// projectValuesModel returns a copy of the model with the static style rules
// removed. Formula and literal cells are left untouched (Tier-1: no compute).
func projectValuesModel(model *Model) *Model {
	next := &Model{Sheets: make([]*Sheet, len(model.Sheets))}
	for i, s := range model.Sheets {
		cp := *s        // shallow copy of the sheet header
		cp.Styles = nil // strip static style rules
		next.Sheets[i] = &cp
	}
	return next
}

// ProjectText renders the values-materialized-for-reading markdown projection,
// minus the compute step (see the TIER-1 BOUNDARY note above). Formula cells
// keep their source text.
func ProjectText(model *Model) string {
	return Serialize(projectValuesModel(model))
}

// ProjectProse renders a flat, prose-like projection: one line per data row,
// `header: value` pairs, prefixed by the sheet name. Ideal as RAG chunk units.
// Formula cells project as their source text (Tier-1: no compute).
func ProjectProse(model *Model) string {
	var lines []string
	for _, sheet := range model.Sheets {
		var headers []string
		if len(sheet.Grid) > 0 {
			headers = sheet.Grid[0]
		}
		for r := 1; r < len(sheet.Grid); r++ {
			var parts []string
			for c := 0; c < sheet.Width; c++ {
				header := ""
				if c < len(headers) {
					header = strings.TrimSpace(headers[c])
				}
				if header == "" {
					header = "Col" + strconv.Itoa(c+1)
				}
				value := GetCell(sheet, c, r+1)
				if strings.TrimSpace(value) != "" {
					parts = append(parts, header+": "+value)
				}
			}
			if len(parts) > 0 {
				lines = append(lines, sheet.Name+" — "+strings.Join(parts, ", "))
			}
		}
	}
	return strings.Join(lines, "\n")
}
